### PR TITLE
feat: Add dependencies.DistributedLockScope

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-416

**Changes:**
- Added distlock.Provider to the dependencies.
- Replaced old/weird Templates API project locker.

-----------

### CODE REVIEW

File: internal/pkg/service/common/dependencies/dependencies.go [hunk 1]
```diff
@@ -137,6 +138,11 @@ type DistributionScope interface {
 	DistributionNode() *distributionPkg.Node
 }
 
+// DistributedLockScope dependencies to acquire distributed locks in the cluster.
+type DistributedLockScope interface {
+	DistributedLockProvider() *distlock.Provider
+}
+
```
Comment (michaljurecko) [inline internal/pkg/service/common/dependencies/dependencies.go:145]: Added `distlock.Provider` to the dependencies.

File: internal/pkg/service/common/distribution/config.go [hunk 2]
```diff
@@ -13,11 +15,12 @@ type Config struct {
 	// All changes in the interval are grouped together, so that updates do not occur too often. Use 0 to disable the grouping.
 	EventsGroupInterval time.Duration `configKey:"eventsGroupInterval" configUsage:"Interval of processing changes in the topology. Use 0 to disable the grouping."  validate:"maxDuration=30s"`
 	// TTLSeconds configures the number seconds after which the node is automatically un-registered if an outage occurs.
-	TTLSeconds int `configKey:"ttlSeconds" configUsage:"Seconds after which the node is automatically un-registered if an outage occurs."  validate:"required,minDuration=1s,maxDuration=30s"`
+	TTLSeconds int `configKey:"ttlSeconds" configUsage:"Seconds after which the node is automatically un-registered if an outage occurs."  validate:"required,min=1,max=30"`
 }
```
Comment (michaljurecko) [inline internal/pkg/service/common/distribution/config.go:19]: Typo

File: internal/pkg/service/common/distribution/node.go [hunk 3]
```diff
@@ -111,12 +111,13 @@ func newGroupMember(nodeID, groupID string, cfg Config, d dependencies) (*GroupN
...
-	_, errCh := etcdop.
+	_, err := etcdop.
 		NewSessionBuilder().
+		WithGrantTimeout(cfg.GrantTimeout).
 		WithTTLSeconds(cfg.TTLSeconds).
 		WithOnSession(g.register).
-		Start(sessionCtx, wg, g.logger, d.EtcdClient())
-	if err := <-errCh; err != nil {
+		StartOrErr(sessionCtx, wg, g.logger, d.EtcdClient())
+	if err != nil {
```
Comment (michaljurecko) [inline internal/pkg/service/common/distribution/node.go:120]: Simplified.

File: internal/pkg/service/templates/api/service/service.go [hunk 4]
```diff
@@ -176,7 +177,7 @@ func (s *service) UseTemplateVersion(ctx context.Context, d dependencies.Project
 	if unlockFn, err := tryLockProject(ctx, d); err != nil {
 		return nil, err
 	} else {
-		defer unlockFn()
+		defer unlockFn(ctx)
 	}
```
Comment (michaljurecko) [inline internal/pkg/service/templates/api/service/service.go:182]: I replaced an old lock mechanism in the Templates API.

File: internal/pkg/service/templates/api/service/service.go [hunk 5]
```diff
@@ -684,21 +685,29 @@ func getTemplateInstance(ctx context.Context, d dependencies.ProjectRequestScope
...
+func tryLockProject(ctx context.Context, d dependencies.ProjectRequestScope) (unlock func(ctx context.Context), err error) {
 	d.Logger().Infof(ctx, `requested lock for project "%d"`, d.ProjectID())
 
-	// Try lock
-	locked, unlockFn := d.ProjectLocker().TryLock(ctx, fmt.Sprintf("project-%d", d.ProjectID()))
-	if !locked {
-		d.Logger().Infof(ctx, `project "%d" is locked by another request`, d.ProjectID())
-		return nil, &ProjectLockedError{
+	mutex := d.DistributedLockProvider().NewMutex(fmt.Sprintf("project/%d", d.ProjectID()))
+
```
Comment (michaljurecko) [inline internal/pkg/service/templates/api/service/service.go:692]: Removed old project locker (working, but implemented in a weird way).

File: internal/pkg/service/common/dependencies/distlock.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+func NewDistributedLockScope(ctx context.Context, cfg distlock.Config, d distributedLockScopeDeps) (DistributedLockScope, error) {
+	return newDistributedLockScope(ctx, cfg, d)
+}
+
+func newDistributedLockScope(ctx context.Context, cfg distlock.Config, d distributedLockScopeDeps) (v *distributedLockScope, err error) {
+	ctx, span := d.Telemetry().Tracer().Start(ctx, "keboola.go.common.dependencies.NewDistributedLockScope")
+	defer span.End(&err)
+
+	provider, err := distlock.NewProvider(cfg, d)
```
Comment (hosekpeter) [inline internal/pkg/service/common/dependencies/distlock.go:28]: I would check for an error.

Comment (michaljurecko) [reply internal/pkg/service/common/dependencies/distlock.go:28]: :+1: thaks, fixed https://github.com/keboola/keboola-as-code/pull/1594/commits/2fe60d96befb393130c630b5a6ead8c5e354b2c3

**Pull request merged**